llvm/lib/Target/AMDGPU/AMDGPUCallLowering.cpp

```diff
@@ -1283,6 +1293,23 @@ bool AMDGPUCallLowering::lowerTailCall(
     MIRBuilder.buildInstr(AMDGPU::ADJCALLSTACKDOWN).addImm(NumBytes).addImm(0);
   }
 
+  // If this is a chain call, we need to set EXEC right before the call.
+  if (AMDGPU::isChainCC(Info.CallConv)) {
+    ArgInfo ExecArg = Info.OrigArgs[1];
+    assert(ExecArg.Regs.size() == 1 && "Too many regs for EXEC");
+
+    if (!ExecArg.Ty->isIntegerTy(ST.getWavefrontSize()))
+      return false;
+
+    unsigned MovOpc = ST.isWave32() ? AMDGPU::S_MOV_B32 : AMDGPU::S_MOV_B64;
+    MCRegister Exec = TRI->getExec();
+    auto SetExec =
+        MIRBuilder.buildInstr(MovOpc).addDef(Exec).addReg(ExecArg.Regs[0]);
```

Comment:
I have some concerns about the exec setup here:
1. Have you checked that a vgpr used in WWM calculation will be restored before the exec setup? Maybe adding a test for this?
2. I feel it would be better we add kind of implicit dependency on `EXEC` to the real call instruction after the exec-setup to explicitly say the call afterward depends on this EXEC setup. This might be helpful to avoid stupid things happening.

Reply:
That seems like a fair point, well spotted. Generally, EXEC modifications are dicey. Not sure, but perhaps this can somehow be integrated into the main call instruction and then expanded later?

Reply:
I think it would be more robust if we view it as kind of one pseudo instruction like `vector_call PC, mask` which got expanded later (maybe at `SILateBranchLowering`?). Can you see other possible issue with doing that?

Reply:
Good catch, I do have a test for this in one of the follow up patches and it's misbehaving. I'll refactor as a pseudo.

I'm not sure about the best place to expand it, is there any reason to go with `SILateBranchLowering` as opposed to `PostRAPseudos` or `AMDGPUMCInstLower`? I'm tempted to go with "as late as possible" because (IMO) it makes it easier to understand things and minimizes the output of `-print-changed`. `AMDGPUMCInstLower` is also where `SI_CALL`, `SI_TCRETURN` etc are expanded, so I'd go with that unless prompted otherwise.

Reply:
I am not familiar with `AMDGPUMCInstLower`, so I don't know whether it is more proper place. If nothing wrong with it, I think it should be ok. I noticed there is still a scheduler run after `PostRAPseudos`. So, it would be better to be later than that, then we don't need to add an implicit use of `EXEC` to the s_setpc, which is a little bit strange.